diff --git a/nixos/modules/installer/tools/get-version-suffix b/nixos/modules/installer/tools/get-version-suffix
index b8972cd57d2..4331c7c3a16 100644
--- a/nixos/modules/installer/tools/get-version-suffix
+++ b/nixos/modules/installer/tools/get-version-suffix
@@ -14,7 +14,7 @@ getVersion() {
     fi
 }
 
-if nixpkgs=$(nix-instantiate --find-file nixpkgs "$@"); then
+if nixpkgs=$(nix-instantiate --find-file nixos "$@"); then
     getVersion $nixpkgs
     if [ -n "$rev" ]; then
         echo ".git.$rev"
diff --git a/nixos/modules/installer/tools/nixos-rebuild.sh b/nixos/modules/installer/tools/nixos-rebuild.sh
index 27e5b5d8c70..45fa7369467 100644
--- a/nixos/modules/installer/tools/nixos-rebuild.sh
+++ b/nixos/modules/installer/tools/nixos-rebuild.sh
@@ -321,7 +321,7 @@ fi
 # Update the version suffix if we're building from Git (so that
 # nixos-version shows something useful).
 if [ -n "$canRun" ]; then
-    if nixpkgs=$(nix-instantiate --find-file nixpkgs "${extraBuildFlags[@]}"); then
+    if nixpkgs=$(nix-instantiate --find-file nixos "${extraBuildFlags[@]}"); then
         suffix=$($SHELL $nixpkgs/nixos/modules/installer/tools/get-version-suffix "${extraBuildFlags[@]}" || true)
         if [ -n "$suffix" ]; then
             echo -n "$suffix" > "$nixpkgs/.version-suffix" || true
